File: frontend/src/types/closing.ts
import type { RespaSection, PaidBy } from "./loan_estimate";

// ── Enums ────────────────────────────────────────────────────────────────────

export enum CDStatus {
  DRAFT = "Draft",
  ISSUED = "Issued",
  REVISED = "Revised",
  SUPERSEDED = "Superseded",
}

export enum ChecklistCategory {
  TITLE = "Title",
  INSURANCE = "Insurance",
  COMPLIANCE = "Compliance",
  FUNDING = "Funding",
  OTHER = "Other",
}

export enum ChecklistItemStatus {
  PENDING = "Pending",
  COMPLETE = "Complete",
  NA = "NA",
}

export enum FundingStatusType {
  NOT_READY = "NotReady",
  SCHEDULED = "Scheduled",
  FUNDED = "Funded",
  SUSPENDED = "Suspended",
}

export const CHECKLIST_CATEGORY_LABELS: Record<ChecklistCategory, string> = {
  [ChecklistCategory.TITLE]: "Title & Settlement",
  [ChecklistCategory.INSURANCE]: "Insurance",
  [ChecklistCategory.COMPLIANCE]: "Compliance",
  [ChecklistCategory.FUNDING]: "Funding",
  [ChecklistCategory.OTHER]: "Other",
};

export const FUNDING_STATUS_LABELS: Record<FundingStatusType, string> = {
  [FundingStatusType.NOT_READY]: "Not Ready",
  [FundingStatusType.SCHEDULED]: "Scheduled",
  [FundingStatusType.FUNDED]: "Funded",
  [FundingStatusType.SUSPENDED]: "Suspended",
};

// ── Closing Disclosure ───────────────────────────────────────────────────────

export interface ClosingDisclosureFee {
  id: string;
  closing_disclosure_id: string;
  respa_section: RespaSection;
  fee_name: string;
  fee_amount: number;
  tolerance_category: string;
  paid_by: PaidBy;
  paid_to?: string;
  is_finance_charge: boolean;
  sort_order: number;
}

export interface ClosingDisclosure {
  id: string;
  loan_id: string;
  version_number: number;
  status: CDStatus;
  issued_date?: string;
  delivery_date?: string;
  closing_date?: string;
  disbursement_date?: string;
  cd_delivery_deadline?: string;
  linked_le_id?: string;
  revision_reason?: string;
  supersedes_cd_id?: string;

  loan_amount: number;
  loan_term_months: number;
  note_rate_percent?: number;
  apr_percent?: number;

  monthly_principal_interest?: number;
  monthly_mortgage_insurance?: number;
  monthly_escrow_amount?: number;
  total_monthly_payment?: number;

  total_loan_costs?: number;
  total_other_costs?: number;
  lender_credits: number;
  total_closing_costs?: number;

  purchase_price?: number;
  down_payment?: number;
  deposit: number;
  seller_credits: number;
  cash_to_close?: number;

  issued_by_id?: string;
  created_at: string;
  updated_at: string;
  fees: ClosingDisclosureFee[];
}

export interface CDListItem {
  id: string;
  loan_id: string;
  version_number: number;
  status: CDStatus;
  issued_date?: string;
  closing_date?: string;
  cd_delivery_deadline?: string;
  note_rate_percent?: number;
  apr_percent?: number;
  total_closing_costs?: number;
  cash_to_close?: number;
  total_monthly_payment?: number;
  created_at: string;
}

export interface CDFeeCreate {
  respa_section: RespaSection;
  fee_name: string;
  fee_amount: number;
  paid_by: PaidBy;
  paid_to?: string;
  is_finance_charge: boolean;
  sort_order?: number;
}

export interface CDCreatePayload {
  loan_term_months: number;
  note_rate_percent?: number;
  monthly_mortgage_insurance?: number;
  monthly_escrow_amount?: number;
  purchase_price?: number;
  deposit: number;
  seller_credits: number;
  lender_credits: number;
  closing_date?: string;
  disbursement_date?: string;
  fees: CDFeeCreate[];
}

export interface CDUpdatePayload {
  note_rate_percent?: number;
  loan_term_months?: number;
  monthly_mortgage_insurance?: number;
  monthly_escrow_amount?: number;
  purchase_price?: number;
  deposit?: number;
  seller_credits?: number;
  lender_credits?: number;
  closing_date?: string;
  disbursement_date?: string;
}

export interface CDIssuePayload {
  issued_date: string;
  delivery_date?: string;
}

// ── Tolerance Check ──────────────────────────────────────────────────────────

export interface ToleranceViolation {
  fee_name: string;
  original: number;
  revised: number;
  increase: number;
}

export interface ToleranceCheckResult {
  zero_violations: ToleranceViolation[];
  ten_pct_bucket_original: number;
  ten_pct_bucket_revised: number;
  ten_pct_exceeded: boolean;
  cure_amount: number;
}

// ── Closing Checklist ────────────────────────────────────────────────────────

export interface ChecklistItem {
  id: string;
  loan_id: string;
  category: ChecklistCategory;
  status: ChecklistItemStatus;
  description: string;
  due_date?: string;
  completed_date?: string;
  assigned_to?: string;
  notes?: string;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface ChecklistItemCreatePayload {
  category: ChecklistCategory;
  description: string;
  due_date?: string;
  assigned_to?: string;
  notes?: string;
  sort_order?: number;
}

export interface ChecklistItemUpdatePayload {
  category?: ChecklistCategory;
  status?: ChecklistItemStatus;
  description?: string;
  due_date?: string;
  completed_date?: string;
  assigned_to?: string;
  notes?: string;
  sort_order?: number;
}

// ── Wire Instructions ────────────────────────────────────────────────────────

export interface WireInstruction {
  id: string;
  loan_id: string;
  beneficiary_name: string;
  beneficiary_address?: string;
  bank_name: string;
  aba_routing_number: string;
  account_number: string;
  reference_number?: string;
  special_instructions?: string;
  verified: boolean;
  verified_by_id?: string;
  created_at: string;
  updated_at: string;
}

export interface WireInstructionPayload {
  beneficiary_name: string;
  beneficiary_address?: string;
  bank_name: string;
  aba_routing_number: string;
  account_number: string;
  reference_number?: string;
  special_instructions?: string;
}

// ── Funding Status ───────────────────────────────────────────────────────────

export interface FundingStatusData {
  id: string;
  loan_id: string;
  status: FundingStatusType;
  scheduled_date?: string;
  funded_date?: string;
  funded_amount?: number;
  confirmation_number?: string;
  funding_source?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface FundingStatusPayload {
  status: FundingStatusType;
  scheduled_date?: string;
  funded_date?: string;
  funded_amount?: number;
  confirmation_number?: string;
  funding_source?: string;
  notes?: string;
}

// ── Default Checklist Items ──────────────────────────────────────────────────

export const DEFAULT_CHECKLIST_ITEMS: Array<{
  category: ChecklistCategory;
  description: string;
}> = [
  { category: ChecklistCategory.TITLE, description: "Title commitment received and reviewed" },
  { category: ChecklistCategory.TITLE, description: "Title exceptions cleared" },
  { category: ChecklistCategory.TITLE, description: "Settlement agent confirmed" },
  { category: ChecklistCategory.TITLE, description: "Closing date/time confirmed with all parties" },
  { category: ChecklistCategory.INSURANCE, description: "Homeowner's insurance binder received" },
  { category: ChecklistCategory.INSURANCE, description: "Flood insurance confirmation (if required)" },
  { category: ChecklistCategory.INSURANCE, description: "Mortgage insurance certificate (if applicable)" },
  { category: ChecklistCategory.COMPLIANCE, description: "Closing Disclosure issued to borrower" },
  { category: ChecklistCategory.COMPLIANCE, description: "3-day CD waiting period satisfied" },
  { category: ChecklistCategory.COMPLIANCE, description: "Final TRID tolerance check — no cure needed" },
  { category: ChecklistCategory.COMPLIANCE, description: "Right of rescission period clear (refi only)" },
  { category: ChecklistCategory.FUNDING, description: "Final underwriting approval (clear to close)" },
  { category: ChecklistCategory.FUNDING, description: "Loan documents drawn and reviewed" },
  { category: ChecklistCategory.FUNDING, description: "Borrower signed closing package" },
  { category: ChecklistCategory.FUNDING, description: "Wire instructions verified" },
  { category: ChecklistCategory.FUNDING, description: "Funding authorization issued" },
];

File: backend/app/schemas/closing.py
from datetime import datetime, date
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, ConfigDict

from app.models.closing import (
    CDStatus,
    ChecklistCategory,
    ChecklistItemStatus,
    FundingStatusType,
)
from app.models.loan_estimate import RespaSection, PaidBy


# ── Closing Disclosure Fees ───────────────────────────────────────────────────


class CDFeeCreate(BaseModel):
    respa_section: RespaSection
    fee_name: str = Field(min_length=1, max_length=200)
    fee_amount: Decimal
    paid_by: PaidBy = PaidBy.BORROWER
    paid_to: Optional[str] = Field(default=None, max_length=200)
    is_finance_charge: bool = True
    sort_order: int = 0


class CDFeeResponse(BaseModel):
    id: UUID
    closing_disclosure_id: UUID
    respa_section: RespaSection
    fee_name: str
    fee_amount: Decimal
    tolerance_category: str
    paid_by: PaidBy
    paid_to: Optional[str] = None
    is_finance_charge: bool
    sort_order: int

    model_config = ConfigDict(from_attributes=True)


# ── Closing Disclosure ────────────────────────────────────────────────────────


class CDCreate(BaseModel):
    loan_term_months: int = 360
    note_rate_percent: Optional[Decimal] = None
    monthly_mortgage_insurance: Optional[Decimal] = Field(default=Decimal("0.00"))
    monthly_escrow_amount: Optional[Decimal] = Field(default=Decimal("0.00"))
    purchase_price: Optional[Decimal] = None
    deposit: Decimal = Decimal("0.00")
    seller_credits: Decimal = Decimal("0.00")
    lender_credits: Decimal = Decimal("0.00")
    closing_date: Optional[str] = Field(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    disbursement_date: Optional[str] = Field(
        default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"
    )
    fees: list[CDFeeCreate] = []


class CDUpdate(BaseModel):
    note_rate_percent: Optional[Decimal] = None
    loan_term_months: Optional[int] = None
    monthly_mortgage_insurance: Optional[Decimal] = None
    monthly_escrow_amount: Optional[Decimal] = None
    purchase_price: Optional[Decimal] = None
    deposit: Optional[Decimal] = None
    seller_credits: Optional[Decimal] = None
    lender_credits: Optional[Decimal] = None
    closing_date: Optional[str] = Field(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    disbursement_date: Optional[str] = Field(
        default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"
    )


class CDIssuePayload(BaseModel):
    issued_date: str = Field(pattern=r"^\d{4}-\d{2}-\d{2}$")
    delivery_date: Optional[str] = Field(
        default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"
    )


class CDResponse(BaseModel):
    id: UUID
    loan_id: UUID
    version_number: int
    status: CDStatus
    issued_date: Optional[date] = None
    delivery_date: Optional[date] = None
    closing_date: Optional[date] = None
    disbursement_date: Optional[date] = None
    cd_delivery_deadline: Optional[date] = None
    linked_le_id: Optional[UUID] = None
    revision_reason: Optional[str] = None
    supersedes_cd_id: Optional[UUID] = None

    loan_amount: Decimal
    loan_term_months: int
    note_rate_percent: Optional[Decimal] = None
    apr_percent: Optional[Decimal] = None

    monthly_principal_interest: Optional[Decimal] = None
    monthly_mortgage_insurance: Optional[Decimal] = None
    monthly_escrow_amount: Optional[Decimal] = None
    total_monthly_payment: Optional[Decimal] = None

    total_loan_costs: Optional[Decimal] = None
    total_other_costs: Optional[Decimal] = None
    lender_credits: Decimal
    total_closing_costs: Optional[Decimal] = None

    purchase_price: Optional[Decimal] = None
    down_payment: Optional[Decimal] = None
    deposit: Decimal
    seller_credits: Decimal
    cash_to_close: Optional[Decimal] = None

    issued_by_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    fees: list[CDFeeResponse] = []

    model_config = ConfigDict(from_attributes=True)


class CDListItem(BaseModel):
    id: UUID
    loan_id: UUID
    version_number: int
    status: CDStatus
    issued_date: Optional[date] = None
    closing_date: Optional[date] = None
    cd_delivery_deadline: Optional[date] = None
    note_rate_percent: Optional[Decimal] = None
    apr_percent: Optional[Decimal] = None
    total_closing_costs: Optional[Decimal] = None
    cash_to_close: Optional[Decimal] = None
    total_monthly_payment: Optional[Decimal] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Tolerance Check ───────────────────────────────────────────────────────────


class ToleranceViolation(BaseModel):
    fee_name: str
    original: float
    revised: float
    increase: float


class ToleranceCheckResponse(BaseModel):
    zero_violations: list[ToleranceViolation]
    ten_pct_bucket_original: float
    ten_pct_bucket_revised: float
    ten_pct_exceeded: bool
    cure_amount: float


# ── Closing Checklist ─────────────────────────────────────────────────────────


class ChecklistItemCreate(BaseModel):
    category: ChecklistCategory
    description: str = Field(min_length=1, max_length=500)
    due_date: Optional[str] = Field(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    assigned_to: Optional[str] = Field(default=None, max_length=100)
    notes: Optional[str] = Field(default=None, max_length=2000)
    sort_order: int = 0


class ChecklistItemUpdate(BaseModel):
    category: Optional[ChecklistCategory] = None
    status: Optional[ChecklistItemStatus] = None
    description: Optional[str] = Field(default=None, min_length=1, max_length=500)
    due_date: Optional[str] = Field(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    completed_date: Optional[str] = Field(
        default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"
    )
    assigned_to: Optional[str] = Field(default=None, max_length=100)
    notes: Optional[str] = Field(default=None, max_length=2000)
    sort_order: Optional[int] = None


class ChecklistItemResponse(BaseModel):
    id: UUID
    loan_id: UUID
    category: ChecklistCategory
    status: ChecklistItemStatus
    description: str
    due_date: Optional[str] = None
    completed_date: Optional[str] = None
    assigned_to: Optional[str] = None
    notes: Optional[str] = None
    sort_order: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Wire Instructions ─────────────────────────────────────────────────────────


class WireInstructionUpsert(BaseModel):
    beneficiary_name: str = Field(min_length=1, max_length=200)
    beneficiary_address: Optional[str] = Field(default=None, max_length=500)
    bank_name: str = Field(min_length=1, max_length=200)
    aba_routing_number: str = Field(min_length=9, max_length=9, pattern=r"^\d{9}$")
    account_number: str = Field(min_length=1, max_length=34)
    reference_number: Optional[str] = Field(default=None, max_length=100)
    special_instructions: Optional[str] = Field(default=None, max_length=2000)


class WireInstructionResponse(BaseModel):
    id: UUID
    loan_id: UUID
    beneficiary_name: str
    beneficiary_address: Optional[str] = None
    bank_name: str
    aba_routing_number: str
    account_number: str
    reference_number: Optional[str] = None
    special_instructions: Optional[str] = None
    verified: bool
    verified_by_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Funding Status ────────────────────────────────────────────────────────────


class FundingStatusUpsert(BaseModel):
    status: FundingStatusType
    scheduled_date: Optional[str] = Field(
        default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"
    )
    funded_date: Optional[str] = Field(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    funded_amount: Optional[Decimal] = None
    confirmation_number: Optional[str] = Field(default=None, max_length=100)
    funding_source: Optional[str] = Field(default=None, max_length=200)
    notes: Optional[str] = Field(default=None, max_length=2000)


class FundingStatusResponse(BaseModel):
    id: UUID
    loan_id: UUID
    status: FundingStatusType
    scheduled_date: Optional[date] = None
    funded_date: Optional[date] = None
    funded_amount: Optional[Decimal] = None
    confirmation_number: Optional[str] = None
    funding_source: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
